### ISSUE TITLE
npm pack fails with directories with special characters

### ISSUE THREAD

**Issue by haraldF** [npm/npm#18328]
#### I'm opening this issue because:

  - [x] npm is crashing.
  - [ ] npm is producing an incorrect install.
  - [ ] npm is doing something I don't understand.
  - [ ] Other (_see below for feature requests_):

#### What's going wrong?

I'm trying to run `npm pack` on one of our packages, but npm crashes. Root cause seems to be that there's a directory containing an `@` sign, which is not handled correctly. (Note - not having `@` in a directory name is not an option).

From the error message, it seems that the `@` sign is stripped from the path, and then an ENOENT occurs.

#### How can the CLI team reproduce the problem?

Create a dummy package in a new directory and add a file in a directory with an '@' character:

```bash
npm init -y
mkdir @foo
touch @foo/test.js
npm pack
```

The output:

```
(node:27087) UnhandledPromiseRejectionWarning: Unhandled promise rejection (rejection id: 1): Error: ENOENT: no such file or directory, stat '/home/harald/tmp/npmtest/foo/test.js'
(node:27087) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.
npm ERR! cb() never called!
```

The npm-debug log can be found here:
https://gist.github.com/haraldF/97f515427f87cc361de2a545573e46c4

### supporting information:

 - `npm -v` prints: 5.4.0
 - `node -v` prints: v8.4.0
 - `npm config get registry` prints: https://registry.npmjs.org/
 - Windows, OS X/macOS, or Linux?: Linux
 - Network issues:
   - Geographic location where npm was run:
   - [ ] I use a proxy to connect to the npm registry.
   - [ ] I use a proxy to connect to the web.
   - [ ] I use a proxy when downloading Git repos.
   - [ ] I access the npm registry via a VPN
   - [ ] I don't use a proxy, but have limited or unreliable internet access.
 - Container:
   - [ ] I develop using Vagrant on Windows.
   - [ ] I develop using Vagrant on OS X or Linux.
   - [ ] I develop / deploy using Docker.
   - [ ] I deploy to a PaaS (Triton, Heroku).

**Comment by haraldF** [comment]
Note - this is a regression, npm 5.3.0 works just fine.

**Comment by haraldF** [comment]
Bug still exists in npm 5.5.x

**Comment by rkoehn** [comment]
It seems the regression was introduced by https://github.com/npm/npm/commit/4cd6a1774f774506323cae5685c9ca9a10deab63 tar@4.0.1

**Comment by zkat** [comment]
Closing as duplicate of https://github.com/npm/npm/issues/18658

There's a pending patch to fix it over at #19033 

Cheers.
